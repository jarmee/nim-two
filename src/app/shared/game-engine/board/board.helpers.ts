import { Board, BoardDifference, BoardDifferences, Column, Columns, Path } from './board.model';

function createBoardDifferencesOf(boardA: Board, andBoardB: Board): BoardDifferences {
  const sourceBoard = boardA || andBoardB;
  return Object.keys(sourceBoard)
    .map(rowKey =>
      Object.keys(sourceBoard[rowKey]).map((columnKey: string) => ({
        newColumn: findColumnOf(boardA, [rowKey, columnKey]),
        currentColumn: findColumnOf(andBoardB, [rowKey, columnKey]),
        path: [rowKey, columnKey]
      }))
    )
    .reduce((flatten: [Column, Column][], element: any) => [...flatten, ...element], []);
}

export function areColumnsDifferentByValue(newColumn: Column, currentColumn: Column): boolean {
  if (!newColumn && !currentColumn) {
    return false;
  }
  if (!newColumn || !currentColumn) {
    return true;
  }
  return newColumn.value !== currentColumn.value;
}

export function findColumnOf(board: Board, path: Path): Column | null {
  if (!board) {
    return null;
  }
  if (!path || !path.length) {
    return null;
  }

  return path.reduce((selectedProperty: any, pathElement: string) => {
    if (!selectedProperty || !selectedProperty[pathElement]) {
      return null;
    }
    return selectedProperty[pathElement];
  }, board);
}

export function differenceOf(boardA: Board, andBoardB: Board): BoardDifferences {
  if (!boardA && !andBoardB) {
    return [];
  }

  return createBoardDifferencesOf(boardA, andBoardB).filter(({ newColumn, currentColumn }: BoardDifference) =>
    areColumnsDifferentByValue(newColumn, currentColumn)
  );
}

export const withColumnValueFalseFilter = (column: Column) => !column.value;

export function countColumnsOf(board: Board, withFilter: (column: Column) => boolean): number {
  if (!board) {
    return 0;
  }

  return Object.values(board)
    .map((columns: Columns) => Object.values(columns).filter(withFilter).length)
    .reduce((amount: number, columnCount: number) => (amount += columnCount), 0);
}

export function exchangeColumnsOf(
  board: Board,
  exchangeColumnCallback: (column: Column, path?: Path) => Column
): Board {
  if (!board) {
    return null;
  }

  return Object.keys(board)
    .map(rowKey => ({
      [rowKey]: Object.keys(board[rowKey])
        .map(columnKey => ({
          [columnKey]: exchangeColumnCallback
            ? exchangeColumnCallback(board[rowKey][columnKey], [rowKey, columnKey])
            : { ...board[rowKey][columnKey] }
        }))
        .reduce((newRow, column) => ({ ...newRow, ...column }))
    }))
    .reduce((newBoard, row) => ({ ...newBoard, ...row }), {});
}
