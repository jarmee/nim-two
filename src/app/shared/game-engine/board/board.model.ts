export enum BoardStatus {
  InProgress = 'IN_PROGRESS'
}
export type Column = {
  value: boolean;
  player?: string;
  errorMessage?: string;
};
export type Columns = Record<number, Column>;
export type Board = {
  [key: number]: Columns;
};
export type Path = string[];
export type BoardDifference = {
  newColumn: Column;
  currentColumn: Column;
  path: Path;
};
export type BoardDifferences = Array<BoardDifference>;
