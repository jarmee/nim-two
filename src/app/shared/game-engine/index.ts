export * from './game-engine.facade';
