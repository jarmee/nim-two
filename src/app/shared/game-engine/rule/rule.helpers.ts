import { cloneDeep } from 'lodash';
import { areColumnsDifferentByValue, exchangeColumnsOf, findColumnOf } from '../board/board.helpers';
import { Board, BoardDifferences, Column, Path } from '../board/board.model';
import { GameState } from '../state/state.model';
import { Player } from '../turn/turn.model';
import { GameRule, GameRules } from './rule.model';

export const applyRules = (rules: GameRules) => (
  newState: GameState,
  actualState: GameState,
  boardDifferences: BoardDifferences
): GameState => {
  if (!rules || !rules.length) return newState;
  return rules.reduce(
    (state: GameState, rule: GameRule) => rule(newState, actualState, boardDifferences)(state),
    cloneDeep(newState)
  );
};

function setPlayerWhereColumnIsDifferent(
  compareToBoard: Board,
  toPlayer: Player
): (column: Column, path: Path) => Column {
  return (column: Column, path: Path) => {
    if (areColumnsDifferentByValue(column, findColumnOf(compareToBoard, path))) {
      return {
        ...column,
        player: toPlayer.name
      };
    }
    return column;
  };
}

export function setPlayerForBoardIn(
  state: GameState,
  onlyToTheChangedColumnsOfState: GameState,
  toPlayer: Player
): GameState {
  if (!state) return null;

  const compareToBoard = onlyToTheChangedColumnsOfState ? onlyToTheChangedColumnsOfState.board : null;
  return {
    ...state,
    board: exchangeColumnsOf(state.board, setPlayerWhereColumnIsDifferent(compareToBoard, toPlayer))
  };
}
