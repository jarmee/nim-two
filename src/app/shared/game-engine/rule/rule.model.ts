import { InjectionToken } from "@angular/core";
import { BoardDifferences } from "../board/board.model";
import { GameState } from "../state/state.model";

export type GameRule = (
  newBoard: GameState,
  currentBoard: GameState,
  differences: BoardDifferences
) => (calculatedGameState: Partial<GameState>) => GameState;

export type GameRules = GameRule[];

export const GAME_RULES: InjectionToken<GameRules> = new InjectionToken<
  GameRules
>("GAME_RULES");
