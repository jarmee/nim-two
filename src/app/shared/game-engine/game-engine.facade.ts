import { Injectable, OnDestroy } from "@angular/core";
import { BehaviorSubject, Observable } from "rxjs";
import { tap } from "rxjs/operators";
import { SubscriptionService } from "../state/subscription.service";
import { Board } from "./board/board.model";
import { BotService } from "./bot/bot.service";
import { RuleService } from "./rule/rule.service";
import { GameState, GameStatus } from "./state/state.model";
import { StateService } from "./state/state.service";
import { Player } from "./turn/turn.model";
import { TurnService } from "./turn/turn.service";

@Injectable()
export class GameEngineFacade extends SubscriptionService implements OnDestroy {
  gameLoop$: BehaviorSubject<Partial<GameState>> = new BehaviorSubject<
    Partial<GameState>
  >(null);

  state$: Observable<GameState> = this.stateService.state$;

  amount$: Observable<number> = this.stateService.amount$;

  board$: Observable<Board> = this.stateService.board$;

  status$: Observable<GameStatus> = this.stateService.status$;

  selectedPlayer$: Observable<Player> = this.turnService.selectedPlayer$;

  constructor(
    private turnService: TurnService,
    private ruleService: RuleService,
    private stateService: StateService,
    private botService: BotService
  ) {
    super();
    this.subscribeTo(
      this.ruleService.rulesApplied$.pipe(
        tap((newState: GameState) => this.stateService.updateState(newState)),
        tap((newState: GameState) => {
          if (newState.status === GameStatus.Valid) {
            this.turnService.switchPlayer();
          }
        })
      )
    );
  }

  executePlay(newBoardState: Board) {
    this.ruleService.applyRules({
      status: GameStatus.Valid,
      board: newBoardState
    });
  }

  reset() {
    this.stateService.resetState();
  }
}
