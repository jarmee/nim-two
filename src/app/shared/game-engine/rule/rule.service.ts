import { Inject, Injectable, Optional } from "@angular/core";
import { BehaviorSubject, Observable } from "rxjs";
import { filter, map, withLatestFrom } from "rxjs/operators";
import { differenceOf } from "../board/board.helpers";
import { BoardDifferences } from "../board/board.model";
import { GameState } from "../state/state.model";
import { StateService } from "../state/state.service";
import { Player } from "../turn/turn.model";
import { TurnService } from "../turn/turn.service";
import { applyRules, setPlayerForBoardIn } from "./rule.helpers";
import { GameRules, GAME_RULES } from "./rule.model";

@Injectable()
export class RuleService {
  applyRules$: BehaviorSubject<Partial<GameState>> = new BehaviorSubject<
    Partial<GameState>
  >(null);

  rulesApplied$: Observable<GameState> = this.applyRules$.pipe(
    filter((state: GameState) => !!state),
    withLatestFrom(this.stateService.state$, this.turnSerivce.selectedPlayer$),
    map(
      ([newGameState, onlyToTheChangedColumnsOfActualState, toPlayer]: [
        GameState,
        GameState,
        Player
      ]) => [
        setPlayerForBoardIn(
          newGameState,
          onlyToTheChangedColumnsOfActualState,
          toPlayer
        ),
        onlyToTheChangedColumnsOfActualState
      ]
    ),
    map(([newGameState, andActualGameState]: [GameState, GameState]) => [
      newGameState,
      andActualGameState,
      differenceOf(newGameState.board, andActualGameState.board)
    ]),
    map(
      ([newState, actualState, boardDifferences]: [
        GameState,
        GameState,
        BoardDifferences
      ]) =>
        applyRules(this.rules)(
          Object.freeze(newState),
          Object.freeze(actualState),
          boardDifferences
        )
    )
  );

  constructor(
    private stateService: StateService,
    private turnSerivce: TurnService,
    @Optional() @Inject(GAME_RULES) private rules: GameRules = []
  ) {}

  applyRules(state: GameState) {
    this.applyRules$.next(state);
  }
}
