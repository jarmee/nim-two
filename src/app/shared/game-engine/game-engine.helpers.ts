import { cloneDeep } from "lodash";
import { areColumnsDifferentByValue as areColumnsDifferentByValue, exchangeColumnsOf, findColumnOf } from "../board/board.helpers";
import { Board, BoardDifferences, Column, Path } from "../board/board.model";
import { AiRule, AiRules, GameRule, GameRules, GameState } from "./game-engine.model";
import { Player } from "./turn/turn.model";

export const applyRules = (rules: GameRules) => (
  newState: GameState,
  actualState: GameState,
  boardDifferences: BoardDifferences
) => {
  if (!rules || !rules.length) return newState;

  return rules.reduce(
    (state: GameState, rule: GameRule) =>
      rule(newState, actualState, boardDifferences)(state),
    cloneDeep(newState)
  );
};

export const calculateState = (rules: AiRules) => (state: GameState) => {
  if (!rules || !rules.length) return state;

  return rules.reduce(
    (calculateState: GameState, rule: AiRule) => rule(state)(calculateState),
    cloneDeep(state)
  );
};

function setPlayerWhereColumnIsDifferent(
  compareToBoard: Board,
  toPlayer: Player
): (column: Column, path: Path) => Column {
  return (column: Column, path: Path) => {
    if (
      areColumnsDifferentByValue(column, findColumnOf(compareToBoard, path))
    ) {
      return {
        ...column,
        player: toPlayer.name
      };
    }
    return column;
  };
}

export function setPlayerForBoardIn(
  state: GameState,
  onlyToTheChangedColumnsOfState: GameState,
  toPlayer: Player
): GameState {
  if (!state) return null;

  const compareToBoard = onlyToTheChangedColumnsOfState
    ? onlyToTheChangedColumnsOfState.board
    : null;
  return {
    ...state,
    board: exchangeColumnsOf(
      state.board,
      setPlayerWhereColumnIsDifferent(compareToBoard, toPlayer)
    )
  };
}
