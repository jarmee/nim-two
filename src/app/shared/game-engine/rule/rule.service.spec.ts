import { TestBed } from "@angular/core/testing";
import { NIM_BOARD } from "../../rules/nim/nim.board";
import { GAME_STATE_STORE } from "../game-engine.model";
import { GameEngineStore } from "../game-engine.store";
import {
  playerFactory,
  PlayerType,
  TURN_STATE_STORE
} from "../turn/turn.model";
import { TurnService } from "../turn/turn.service";
import { TurnStore } from "../turn/turn.store";
import { RuleService } from "./rule.service";

describe("RuleService", () => {
  const PLAYERS = [
    playerFactory("ðŸ˜Ž", PlayerType.Human),
    playerFactory("ðŸ¤–", PlayerType.Artificial)
  ];
  beforeEach(() =>
    TestBed.configureTestingModule({
      providers: [
        RuleService,
        {
          provide: GAME_STATE_STORE,
          useFactory: () => new GameEngineStore(NIM_BOARD)
        },
        {
          provide: TURN_STATE_STORE,
          useFactory: () => new TurnStore(PLAYERS)
        },
        TurnService
      ]
    })
  );

  it("should be created", () => {
    const service: RuleService = TestBed.get(RuleService);
    expect(service).toBeTruthy();
  });
});
