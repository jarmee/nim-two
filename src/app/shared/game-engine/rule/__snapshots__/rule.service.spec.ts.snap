// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleService rulesApplied$ should emit a new game state which matches the snapshot 1`] = `
Object {
  "board": Object {
    "0": Object {
      "0": Object {
        "errorMessage": null,
        "player": "ðŸ˜Ž",
        "value": true,
      },
      "1": Object {
        "errorMessage": null,
        "player": "ðŸ˜Ž",
        "value": true,
      },
      "10": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "11": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "12": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "2": Object {
        "errorMessage": null,
        "player": "ðŸ˜Ž",
        "value": true,
      },
      "3": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "4": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "5": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "6": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "7": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "8": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
      "9": Object {
        "errorMessage": null,
        "player": null,
        "value": false,
      },
    },
  },
  "status": "Valid",
}
`;
